src/types/mod.rs

```diff
@@ -27,6 +27,7 @@ pub use self::string::{PyString, PyString as PyUnicode};
 pub use self::tuple::PyTuple;
 pub use self::typeobject::PyType;
 
+// NOTE: this is used by rust-numpy and takes `$type_param`.
```

Comment:
I must confess I am cautious of giving `rust-numpy` special treatment in the PyO3 codebase. We should definitely always make things possible for `rust-numpy` to work but I don't think we should be special-casing PyO3 code for `rust-numpy`. What if someone wants to work on e.g. `rust-tensorflow` using PyO3 and needs us to make further special cases? Also what if the macros used by `rust-numpy` change in future - we'd then need to make another PyO3 release to change these macros.

I think it's best if we make sure that PyO3 APIs are flexible enough to support all crates without needing to introduce special rules for a crate to work.

So I think really all of these macros should continue to take `$type_param` for best consistency.